### ISSUE TITLE
HttpTimerMetrics with Future

### ISSUE THREAD

**Issue by mazorigal** [Backline/akka-http-metrics#13]
Hi,

I have the following code for defining my route:

```
  val userProfileRoutes = {
    logRequestResult("akka-http-user-profile-service") {
      (get & path("raw" / Segment / Segment)) { (uid, uidType) =>
        timerDirectiveWithName(name(metricBaseName.name, "inbound", "raw", "handle")) {
          responseCodeMetricsWithName(name(metricBaseName.name, "inbound", "raw", "handle"))(executor) {
            complete {
              userProfileApp.getUserProfileRaw(uid, uidType).map[ToResponseMarshallable] {
                case Right(userProfile) => userProfile
                case Left(err) => err match {
                  case errType: UserProfileNotFoundError => NotFound -> CustomErrorPayload(errType)
                  case errType: UserProfileError => InternalServerError -> CustomErrorPayload(errType)
                }
              }
            }
          }
        }
      }
    }
  }
```

the ```userProfileApp.getUserProfileRaw(uid, uidType).map[ToResponseMarshallable]``` is the part that reads data from the DB and transform it. It also returns a Future.

In my Graphite metrics, the mean time for the requests handle (for that route) is significantly lower then the mean time for the data retrieval and transformation from the DB, which obviously make no sense. 

Could it be related to the fact that you are using the simple Timer and not the timeFuture ?

Thanks,

**Comment by mazorigal** [comment]
small update.
I have created the following to see if the Future is indeed related to the timing issue.


```
private[this] val pingTimerTest = metrics.timer(name("ping", "test"))

  def doPing(): Future[String] = {
    pingTimerTest.timeFuture{
      Future{
        Thread.sleep(3000)
        "pong"
      }
    }
  }

  val pingRoutes =
      (get & path("ping")) {
        timerDirectiveWithName(name(metricBaseName.name, "inbound", "ping", "handle")) {
          responseCodeMetricsWithName(name(metricBaseName.name, "inbound", "ping", "handle"))(executor) {
            complete {
              doPing()
            }
          }
        }
      }
    
```

and I still see in Graphite that mean time for request handle is almost 0 whether for the doPing is 3 seconds.

If I define the doPing as following:

```
def doPing(): String = {
    pingTimerTest.time{
        Thread.sleep(3000)
        "pong"
    }
  }
```

Then all the metrics are fine, the mean time of the request handle is almost same as of the doPing function.

**Comment by adamdecaf** [comment]
Yea. The timings for futures won't be recorded properly the way things are now. That's because the instance of `Future` is created pretty quickly, but the computation result is not available for some time after. 

I see the metrics-scala has a `timeFuture` method: https://github.com/erikvanoosten/metrics-scala/blob/master/src/main/scala/nl/grons/metrics/scala/Timer.scala#L90 is that what you meant instead of using a simple dropwizard Timer?

**Comment by mazorigal** [comment]
Yes, that what I indeed meant.
But I think that timeFuture would not work with 
mapInnerRoute, maybe mapRouteResultFuture can help
http://doc.akka.io/docs/akka-http/current/scala/http/routing-dsl/directives/basic-directives/mapRouteResultFuture.html#maprouteresultfuture

**Comment by adamdecaf** [comment]
I'm not quite sure this will work with the methods akka-http exposes. It seems `RouteResult` is a sealed type that's only created after the response is generated (aka the database call is already done. I think we'll need a better hook (something like `mapInnerResultFuture` that would wrap the innerRoute in a future we could call `onComplete` on).

http://doc.akka.io/docs/akka-http/current/scala/http/routing-dsl/routes.html#routeresult
https://github.com/Backline/akka-http-metrics/commit/8dc8d59a216d7154989bb8fda1323cc1ec218d86

**Comment by mazorigal** [comment]
I think that for the moment such method in akka http DSL do not exist.
Any other ideas ?

**Comment by mazorigal** [comment]
maybe that can help:

I was using ```logRequestResult```  from ```DebuggingDirectives``` which comes from the akka-http dsl and logs the response to each request. 

this is how it looks:

```
def logRequestResult(magnet: LoggingMagnet[HttpRequest ⇒ RouteResult ⇒ Unit]): Directive0 =
    extractRequestContext.flatMap { ctx ⇒
      val logResult = magnet.f(ctx.log)(ctx.request)
      mapRouteResult { result ⇒
        logResult(result)
        result
      }
    }
```

I modified it a bit to:

```
private[this] val testTimer = metricRegistry.timer(name("ping", "test"))


  def logRequestResultTimer(): Directive0 =
    extractRequestContext.flatMap { ctx ⇒
      val timer = testTimer.time()
      println("1"*50)
      mapRouteResult { result ⇒
        println("2"*50, timer.stop())
        result
      }
    }
```

worked fine with Futures as well.

what do you think ? could it help ?

**Comment by adamdecaf** [comment]
@mazorigal That could work! Could you try it out with the test I wrote? It's over on commit 8dc8d59a216d7154989bb8fda1323cc1ec218d86

**Comment by mazorigal** [comment]
I tried to test extractRequestContext with Future, from some reason it fails for the Future test.
But I dont really understand what is the test you are trying to have.

what you try to test with the following ?

```
val durationFactor: Double = 1.0D / TimeUnit.SECONDS.toNanos(1)

    def mean: Double = counts.getSnapshot().getMean() * durationFactor
    mean must be_>=(0.09).eventually // > 90ms
```

anyway, I added some console printing to the test, in order to check the mean of the timer, and although the test for the Future fails, the result is as excepted. Maybe you miss something in the test.

here is my gist for what I changed in HttpTimerMetrics:
https://gist.github.com/mazorigal/c894054738eeea3391e6e19b37df15be

and that is the gist for what I added to the test file:
https://gist.github.com/mazorigal/35e022939a2f31a4c84d9eac389fe8be

Please have a look.

**Comment by adamdecaf** [comment]
Awesome! I think that's what we need to do then. My test was trying to get the timing out of each `Snapshot` to make sure the timing was including the waits.

Do you want to make a PR with the change? I can if you want.

**Comment by mazorigal** [comment]
Would be great if you could apply the needed changes.

**Comment by adamdecaf** [comment]
Sure thing. I've created https://github.com/Backline/akka-http-metrics/pull/14 and will release a new version once that passes. 

Thanks!

**Comment by adamdecaf** [comment]
I released 0.4.2. Could you try it out?

https://github.com/Backline/akka-http-metrics/releases/tag/0.4.2

**Comment by mazorigal** [comment]
looks good :-)
